ethexe/processor/src/handling/run.rs

```diff
@@ -16,178 +16,282 @@
+//!   4. Repeat steps 1-3 until no messages are processed, or we run out of processing time/gas allowance.
+//!
+//! ---
+//!
+//! ## Simplest Chunk Splitting Algorithm
+//!
+//! A basic chunk-splitting algorithm is implemented as follows:
+//!
+//!   1. First, calculate a temporary chunk index based on the program queue size.
+//!   2. Next, store the required data in a temporary chunk list.
+//!   3. Repeat this process for all programs with non-empty queues.
+//!   4. Since the number of elements in each temporary chunk is random, they need to be redistributed
+//!      to ensure all final chunks contain an equal number of elements.
+//!      To achieve this, we first merge all temporary chunk lists sequentially
+//!      and then redistribute the data according to the expected chunk size.
+//!   5. Reverse the order of the chunks to ensure that the first chunk contains the largest queue size.
+//!   5. Finally, we return the final chunk list.
+//!
+//! ---
+//!
+//! ## Future Improvements
+//!
+//! Currently, the chunk partitioning algorithm is simple and does not consider a program’s execution time.
+//! In the future, we could introduce a weight multiplier to the queue size to improve partitioning efficiency.
+//! This weight multiplier could be calculated based on program execution time statistics.
+
+use ethexe_common::{db::CodesStorage, StateHashWithQueueSize};
 use ethexe_db::Database;
-use ethexe_runtime_common::{InBlockTransitions, JournalHandler, TransitionController};
-use gear_core::ids::ProgramId;
-use gprimitives::H256;
-use std::collections::BTreeMap;
-use tokio::sync::{mpsc, oneshot};
-
-enum Task {
-    Run {
-        program_id: ProgramId,
-        state_hash: H256,
-        result_sender: oneshot::Sender<(Vec<JournalNote>, Option<Origin>)>,
-    },
-}
+use ethexe_runtime_common::{
+    InBlockTransitions, JournalHandler, ProgramJournals, TransitionController,
+};
+use gprimitives::{ActorId, H256};
+use itertools::Itertools;
+use tokio::task::JoinSet;
+
+use crate::host::{InstanceCreator, InstanceWrapper};
 
-pub fn run(
-    config: &ProcessorConfig,
+pub async fn run(
+    chunk_processing_threads: usize,
     db: Database,
     instance_creator: InstanceCreator,
     in_block_transitions: &mut InBlockTransitions,
 ) {
-    tokio::task::block_in_place(|| {
-        let mut rt_builder = tokio::runtime::Builder::new_multi_thread();
+    let mut join_set = JoinSet::new();
+    let chunk_size = chunk_processing_threads;
 
-        if let Some(worker_threads) = config.worker_threads_override {
-            rt_builder.worker_threads(worker_threads);
-        };
+    loop {
+        let states: Vec<_> = in_block_transitions
+            .states_iter()
+            .filter_map(|(&actor_id, state)| {
+                if state.cached_queue_size == 0 {
+                    return None;
+                }
 
-        rt_builder.enable_all();
+                Some((actor_id, state.clone()))
```

Comment:
`StateHashWithQueueSize` can just implement `Copy`
```suggestion
                Some((actor_id, state))
```

Reply:
Fixed